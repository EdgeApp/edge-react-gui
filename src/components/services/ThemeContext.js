// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
