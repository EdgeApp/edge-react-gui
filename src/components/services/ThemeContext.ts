import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
