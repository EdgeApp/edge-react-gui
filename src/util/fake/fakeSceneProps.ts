import { AppParamList, EdgeSceneProps, NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener() {
    return () => undefined
  },
  removeListener() {},
  isFocused() {
    return true
  },

  navigate() {},
  push() {},
  replace() {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  dispatch() {},
  reset() {},
  canGoBack() {
    return true
  },
  getId() {
    return undefined
  },
  getState() {
    return {
      key: 'foo',
      index: 0,
      routes: [],
      routeNames: [],
      type: 'bar',
      stale: false
    }
  },
  getParent() {
    throw new Error('not implemented')
  },
  setOptions() {}
}

export function fakeSceneProps<Name extends keyof AppParamList>(name: Name, params: AppParamList[Name]): EdgeSceneProps<Name> {
  return {
    navigation: fakeNavigation,
    route: {
      name: name,
      key: name + '-0',
      params
    } as any
  }
}
