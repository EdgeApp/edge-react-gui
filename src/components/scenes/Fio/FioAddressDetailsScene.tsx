import { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { Alert } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import { lstrings } from '../../../locales/strings'
import { connect } from '../../../types/reactRedux'
import { EdgeSceneProps } from '../../../types/routerTypes'
import { BUNDLED_TXS_AMOUNT_ALERT, findWalletByFioAddress } from '../../../util/FioAddressUtils'
import { SceneWrapper } from '../../common/SceneWrapper'
import { ConnectWalletsConnector as ConnectWallets } from '../../FioAddress/ConnectWallets'
import { showError } from '../../services/AirshipInstance'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../../services/ThemeContext'
import { SettingsHeaderRow } from '../../settings/SettingsHeaderRow'
import { SettingsTappableRow } from '../../settings/SettingsTappableRow'
import { EdgeText } from '../../themed/EdgeText'

interface StateProps {
  fioWallets: EdgeCurrencyWallet[]
}

interface OwnProps extends EdgeSceneProps<'fioAddressDetails'> {}

interface LocalState {
  fioWalletLoading: boolean
  fioWallet: EdgeCurrencyWallet | null
}

type Props = StateProps & ThemeProps & OwnProps

export class FioAddressDetails extends React.Component<Props, LocalState> {
  state: LocalState = {
    fioWalletLoading: false,
    fioWallet: null
  }

  componentDidMount() {
    const { route } = this.props
    const { fioAddressName } = route.params
    if (!fioAddressName) {
      Alert.alert(lstrings.fio_address_details_screen_alert_title, lstrings.fio_address_details_screen_alert_message, [
        { text: lstrings.fio_address_details_screen_alert_button }
      ])
    }
    this.findFioWallet().catch(err => showError(err))
  }

  findFioWallet = async () => {
    const { fioWallets, route } = this.props
    const { fioAddressName } = route.params

    this.setState({ fioWalletLoading: true })
    const fioWallet = await findWalletByFioAddress(fioWallets, fioAddressName)
    this.setState({ fioWalletLoading: false, fioWallet })
  }

  _onPressAccountSettings = (): void => {
    const { navigation, route } = this.props
    const { fioAddressName, bundledTxs } = route.params
    const { fioWallet } = this.state
    if (fioWallet) {
      navigation.navigate('fioAddressSettings', {
        fioWallet,
        fioAddressName,
        bundledTxs,
        refreshAfterAddBundledTxs: true
      })
    } else {
      showError(lstrings.fio_wallet_missing_for_fio_address)
    }
  }

  renderAccountSettings = () => {
    const {
      theme,
      route: {
        params: { bundledTxs }
      }
    } = this.props
    const styles = getStyles(theme)

    if (bundledTxs < BUNDLED_TXS_AMOUNT_ALERT) {
      return (
        <SettingsTappableRow onPress={this._onPressAccountSettings}>
          <IonIcon name="ios-warning" color={theme.warningIcon} style={styles.settingsIcon} />
          <EdgeText style={styles.settingsWarning} numberOfLines={4}>
            {!bundledTxs ? lstrings.fio_address_details_no_bundled_txs : lstrings.fio_address_details_bundled_txs_out_soon}
          </EdgeText>
        </SettingsTappableRow>
      )
    }

    return (
      <SettingsTappableRow label={lstrings.fio_address_details_screen_manage_account_settings} onPress={this._onPressAccountSettings}>
        <IonIcon name="ios-settings" color={theme.icon} style={styles.settingsIcon} />
      </SettingsTappableRow>
    )
  }

  render() {
    const { navigation, theme, route } = this.props
    const { fioAddressName, bundledTxs } = route.params
    const styles = getStyles(theme)
    const bundledTxsLabel = `${lstrings.fio_address_details_screen_bundled_txs}: ${bundledTxs}`

    return (
      <SceneWrapper background="theme">
        <EdgeText style={styles.bundledTxs}>{bundledTxsLabel}</EdgeText>
        {this.renderAccountSettings()}
        <SettingsHeaderRow
          icon={<IonIcon name="ios-link" color={theme.primaryText} size={theme.rem(1.5)} />}
          label={lstrings.fio_address_details_connect_to_wallets}
        />
        <ConnectWallets fioAddressName={fioAddressName} fioWallet={this.state.fioWallet} navigation={navigation} disabled={this.state.fioWalletLoading} />
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  text: {
    color: theme.primaryText,
    fontSize: theme.rem(1)
  },
  bundledTxs: {
    fontSize: theme.rem(0.75),
    color: theme.primaryText,
    textAlign: 'center',
    marginTop: theme.rem(-0.5),
    paddingBottom: theme.rem(0.75)
  },
  settingsIcon: {
    fontSize: theme.rem(1.5),
    paddingHorizontal: theme.rem(0.5)
  },
  settingsWarning: {
    fontSize: theme.rem(0.75),
    color: theme.warningText,
    paddingLeft: theme.rem(0.5),
    width: '75%'
  }
}))

export const FioAddressDetailsScene = connect<StateProps, {}, OwnProps>(
  state => ({
    fioWallets: state.ui.wallets.fioWallets
  }),
  dispatch => ({})
)(withTheme(FioAddressDetails))
