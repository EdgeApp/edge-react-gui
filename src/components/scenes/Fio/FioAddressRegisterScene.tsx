import { EdgeCurrencyConfig, EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { Image, LayoutChangeEvent, ScrollView, View } from 'react-native'
import { sprintf } from 'sprintf-js'

import { createFioWallet } from '../../../actions/FioAddressActions'
import { Fontello } from '../../../assets/vector'
import { FIO_ADDRESS_DELIMITER, FIO_DOMAIN_DEFAULT } from '../../../constants/WalletAndCurrencyConstants'
import { lstrings } from '../../../locales/strings'
import { connect } from '../../../types/reactRedux'
import { EdgeSceneProps } from '../../../types/routerTypes'
import { FioDomain, FioPublicDomain } from '../../../types/types'
import { getWalletName } from '../../../util/CurrencyWalletHelpers'
import { checkIsDomainPublic } from '../../../util/FioAddressUtils'
import { openLink } from '../../../util/utils'
import { SceneWrapper } from '../../common/SceneWrapper'
import { DomainListModal } from '../../FioAddress/DomainListModal'
import { TextInputModal } from '../../modals/TextInputModal'
import { WalletListModal, WalletListResult } from '../../modals/WalletListModal'
import { Airship, showError, showToast } from '../../services/AirshipInstance'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../../services/ThemeContext'
import { ClickableText } from '../../themed/ClickableText'
import { EdgeText } from '../../themed/EdgeText'
import { FormError } from '../../themed/FormError'
import { MainButton } from '../../themed/MainButton'
import { SceneHeader } from '../../themed/SceneHeader'
import { Tile } from '../../tiles/Tile'

interface State {
  selectedWallet: EdgeCurrencyWallet | null
  selectedDomain: FioDomain
  publicDomains: FioDomain[]
  fioAddress: string
  isValid: boolean
  loading: boolean
  walletLoading: boolean
  domainsLoading: boolean
  isAvailable: boolean
  fieldPos: number
  inputWidth: number
  showFreeAddressLink: boolean
  errorMessage: string
}

interface OwnProps extends EdgeSceneProps<'fioAddressRegister'> {}

interface StateProps {
  fioWallets: EdgeCurrencyWallet[]
  fioPlugin: EdgeCurrencyConfig
  isConnected: boolean
}

interface DispatchProps {
  createFioWallet: () => Promise<EdgeCurrencyWallet>
}

type Props = StateProps & DispatchProps & OwnProps & ThemeProps

export class FioAddressRegister extends React.Component<Props, State> {
  fioCheckQueue: number = 0

  state: State = {
    selectedWallet: null,
    selectedDomain: FIO_DOMAIN_DEFAULT,
    publicDomains: [],
    fioAddress: '',
    isValid: true,
    isAvailable: false,
    loading: false,
    walletLoading: false,
    domainsLoading: true,
    fieldPos: 200,
    inputWidth: 0,
    showFreeAddressLink: false,
    errorMessage: ''
  }

  componentDidMount() {
    const { fioWallets } = this.props
    this.getPublicDomains().catch(err => showError(err))
    this.checkFreeAddress().catch(err => showError(err))
    if (fioWallets.length > 0) {
      this.setState({
        selectedWallet: fioWallets[0]
      })
    } else {
      this.createFioWallet().catch(err => showError(err))
    }
    this.setState({ inputWidth: this.props.theme.rem(12.5) })
  }

  checkFreeAddress = async () => {
    try {
      const { fioPlugin } = this.props
      const publicDomains = await fioPlugin.otherMethods.getDomains(fioPlugin.currencyInfo.defaultSettings.freeAddressRef)
      if (publicDomains.findIndex((publicDomain: FioPublicDomain) => publicDomain.free) > -1) {
        this.setState({ showFreeAddressLink: true })
      }
    } catch (e: any) {
      //
      console.log(e)
    }
  }

  getPublicDomains = async () => {
    const { fioPlugin } = this.props
    try {
      const publicDomains = await fioPlugin.otherMethods.getDomains(fioPlugin.currencyInfo.defaultSettings.fallbackRef)
      const publicDomainsConverted = publicDomains
        // @ts-expect-error
        .sort(publicDomain => (publicDomain.domain === FIO_DOMAIN_DEFAULT.name ? -1 : 1))
        .map((publicDomain: FioPublicDomain) => ({
          name: publicDomain.domain,
          expiration: new Date().toDateString(),
          isPublic: true,
          walletId: '',
          isFree: publicDomain.free
        }))
      this.setState({
        publicDomains: publicDomainsConverted,
        selectedDomain: publicDomainsConverted[0]
      })
    } catch (e: any) {
      //
    }
    this.setState({ domainsLoading: false })
  }

  createFioWallet = async (): Promise<void> => {
    const { createFioWallet } = this.props
    showToast(lstrings.preparing_fio_wallet)
    this.setState({ walletLoading: true })
    try {
      const wallet = await createFioWallet()
      this.setState({
        selectedWallet: wallet,
        walletLoading: false
      })
    } catch (e: any) {
      this.setState({ walletLoading: false })
      showError(lstrings.create_wallet_failed_message)
    }
  }

  registerFreeAddress = async () => {
    const { fioPlugin, fioWallets } = this.props
    const { selectedWallet } = this.state
    if (!fioPlugin) return
    if (!fioWallets.length) return
    if (!selectedWallet) return
    const publicKey = selectedWallet.publicWalletInfo.keys.publicKey
    const url = `${fioPlugin.currencyInfo.defaultSettings.fioAddressRegUrl}${fioPlugin.currencyInfo.defaultSettings.freeAddressRef}?publicKey=${publicKey}`
    try {
      await openLink(url)
    } catch (e: any) {
      showError(sprintf(lstrings.open_url_err, url))
    }
  }

  handleNextButton = (): void => {
    const { isConnected, navigation } = this.props
    const { fioAddress, selectedWallet, selectedDomain, isValid, isAvailable, loading, walletLoading } = this.state
    if (isValid && isAvailable && !loading && !walletLoading) {
      if (isConnected) {
        if (!selectedWallet) return showError(lstrings.create_wallet_failed_message)
        const fullAddress = `${fioAddress}${FIO_ADDRESS_DELIMITER}${selectedDomain.name}`
        if (selectedDomain.isFree) {
          navigation.navigate('fioNameConfirm', {
            fioName: fullAddress,
            paymentWallet: selectedWallet,
            fee: 0,
            ownerPublicKey: selectedWallet.publicWalletInfo.keys.publicKey
          })
        } else {
          navigation.navigate('fioAddressRegisterSelectWallet', {
            fioAddress: fullAddress,
            selectedWallet,
            selectedDomain
          })
        }
      } else {
        showError(lstrings.fio_network_alert_text)
      }
    }
  }

  checkFioAddress(fioAddress: string, domain: string, isCustomDomain: boolean = false) {
    this.setState({
      loading: true,
      errorMessage: ''
    })
    this.fioCheckQueue++
    setTimeout(async () => {
      // do not check if user continue typing fio address
      if (this.fioCheckQueue > 1) {
        return --this.fioCheckQueue
      }
      this.fioCheckQueue = 0
      const { fioPlugin } = this.props
      if (isCustomDomain) {
        try {
          await checkIsDomainPublic(fioPlugin, domain)
        } catch (e: any) {
          showError(e.message)
          return this.setState({
            isAvailable: false,
            loading: false
          })
        }
      }

      // regexp from edge-currency-accountbased
      if (!/^[a-zA-Z0-9]{1}((?!-{2,})[a-zA-Z0-9-]*[a-zA-Z0-9]+){0,1}$/.test(fioAddress)) {
        this.setState({
          loading: false,
          isValid: false,
          errorMessage: lstrings.warning_alphanumeric
        })
        return
      }

      try {
        const fullAddress = `${fioAddress}${FIO_ADDRESS_DELIMITER}${domain}`
        const isAvailable = fioPlugin.otherMethods ? await fioPlugin.otherMethods.validateAccount(fullAddress) : false
        this.setState({
          isValid: true,
          isAvailable,
          loading: false,
          errorMessage: ''
        })
      } catch (e: any) {
        console.log(e.json)
        console.log(e.message)
        this.setState({
          isValid: false,
          isAvailable: false,
          loading: false
        })
      }
    }, 1000)
  }

  handleFioAddressChange = (fioAddressChanged: string) => {
    if (!this.props.isConnected) {
      return this.setState({
        fioAddress: fioAddressChanged.toLowerCase(),
        loading: false
      })
    }
    this.checkFioAddress(fioAddressChanged, this.state.selectedDomain.name, !this.state.selectedDomain.walletId)

    this.setState({
      fioAddress: fioAddressChanged.toLowerCase()
    })
  }

  handleFioAddressFocus = () => {
    // @ts-expect-error
    // eslint-disable-next-line react/no-string-refs
    this.refs._scrollView.scrollTo({ x: 0, y: this.state.fieldPos, animated: true })
  }

  fieldViewOnLayout = ({ nativeEvent }: LayoutChangeEvent) => {
    if (nativeEvent) {
      const {
        layout: { y }
      } = nativeEvent
      this.setState({ fieldPos: y })
    }
  }

  editAddressPressed = async () => {
    this.handleFioAddressFocus()
    await Airship.show<string | undefined>(bridge => (
      <TextInputModal
        bridge={bridge}
        initialValue={this.state.fioAddress}
        inputLabel={lstrings.fio_address_register_form_field_label}
        title={lstrings.fio_address_choose_label}
      />
    )).then((response: string | undefined) => {
      if (response) {
        this.handleFioAddressChange(response)
      }
    })
  }

  handleFioWalletChange = (walletId: string) => {
    this.setState({
      // @ts-expect-error
      selectedWallet: this.props.fioWallets.find(fioWallet => fioWallet.id === walletId)
    })
  }

  selectFioWallet = async () => {
    await Airship.show<WalletListResult>(bridge => (
      <WalletListModal bridge={bridge} navigation={this.props.navigation} headerTitle={lstrings.select_wallet} allowedAssets={[{ pluginId: 'fio' }]} />
    )).then(({ walletId, currencyCode }: WalletListResult) => {
      if (walletId && currencyCode) {
        this.handleFioWalletChange(walletId)
      }
    })
  }

  selectFioDomain = async () => {
    const { navigation } = this.props
    const { domainsLoading } = this.state
    if (domainsLoading) return
    const response = await Airship.show<FioDomain | undefined>(bridge => (
      <DomainListModal bridge={bridge} navigation={navigation} publicDomains={this.state.publicDomains} />
    ))
    if (response) {
      this.setState({ selectedDomain: response })
      this.checkFioAddress(this.state.fioAddress, response.name, !response.walletId)
    }
  }

  renderButton() {
    const { isValid, isAvailable, loading, walletLoading } = this.state
    const styles = getStyles(this.props.theme)

    if (isValid && isAvailable && !loading) {
      return (
        <View style={styles.buttons}>
          <MainButton
            disabled={!isAvailable || walletLoading}
            label={walletLoading ? '' : lstrings.string_next_capitalized}
            spinner={walletLoading}
            onPress={this.handleNextButton}
            type="secondary"
          />
        </View>
      )
    }

    return null
  }

  renderLoader() {
    const { theme } = this.props
    const { loading } = this.state
    const styles = getStyles(theme)

    const label = `(${lstrings.validating})`
    return loading ? <EdgeText style={styles.muted}>{label}</EdgeText> : null
  }

  renderFioWallets() {
    const { fioWallets } = this.props
    const { selectedWallet } = this.state

    if (fioWallets && fioWallets.length > 1) {
      const title = `${selectedWallet == null ? lstrings.fio_address_register_no_wallet_name : getWalletName(selectedWallet)}`
      return <Tile type="touchable" title={`${lstrings.title_fio_connect_to_wallet}`} onPress={this.selectFioWallet} body={title} />
    }
  }

  renderErrorMessage() {
    const { fioAddress, isAvailable, isValid, loading, errorMessage } = this.state
    const styles = getStyles(this.props.theme)
    let chooseHandleErrorMessage = ''

    if (loading) return null

    if (fioAddress && !this.props.isConnected) {
      chooseHandleErrorMessage = lstrings.fio_address_register_screen_cant_check
    }
    if (fioAddress && !isAvailable) {
      chooseHandleErrorMessage = lstrings.fio_address_register_screen_not_available
    }

    if (fioAddress && !isValid) {
      chooseHandleErrorMessage = lstrings.fio_error_invalid_address
    }

    if (fioAddress && !isValid && errorMessage) {
      chooseHandleErrorMessage = errorMessage
    }

    return (
      <FormError style={styles.error} isVisible={!!chooseHandleErrorMessage}>
        {chooseHandleErrorMessage}
      </FormError>
    )
  }

  render() {
    const { theme } = this.props
    const { fioAddress, selectedDomain, domainsLoading, showFreeAddressLink } = this.state
    const styles = getStyles(theme)

    return (
      <SceneWrapper background="theme">
        <SceneHeader style={styles.header} title={lstrings.title_fio_address_confirmation}>
          <Image source={theme.fioAddressLogo} style={styles.image} resizeMode="cover" />
        </SceneHeader>
        {/* eslint-disable-next-line react/no-string-refs */}
        <ScrollView ref="_scrollView" contentContainerStyle={styles.container}>
          <View style={styles.view}>
            <View style={[styles.createWalletPromptArea, styles.title]}>
              <EdgeText style={styles.instructionalText} numberOfLines={2}>
                {lstrings.fio_address_first_screen_title}
              </EdgeText>
            </View>
            <View style={styles.createWalletPromptArea}>
              <EdgeText style={styles.handleRequirementsText} numberOfLines={3}>
                {lstrings.fio_address_features}
              </EdgeText>
            </View>
            <View style={styles.createWalletPromptArea}>
              <EdgeText style={styles.handleRequirementsText} numberOfLines={5}>
                {lstrings.fio_address_first_screen_end}
              </EdgeText>
            </View>

            <View onLayout={this.fieldViewOnLayout}>
              <Tile type="editable" title={lstrings.fio_address_choose_label} onPress={this.editAddressPressed}>
                <View style={styles.addressTileBody}>
                  {fioAddress ? (
                    <EdgeText style={styles.fioAddressName}>{fioAddress}</EdgeText>
                  ) : (
                    <EdgeText style={styles.muted}>{lstrings.fio_address_register_placeholder}</EdgeText>
                  )}
                  {this.renderLoader()}
                </View>
              </Tile>
              <Tile
                type="touchable"
                title={lstrings.fio_address_choose_domain_label}
                onPress={this.selectFioDomain}
                body={domainsLoading ? lstrings.loading : `${FIO_ADDRESS_DELIMITER}${selectedDomain.name}`}
              />
              {this.renderFioWallets()}
            </View>
            {this.renderButton()}
            {this.props.fioWallets.length && showFreeAddressLink ? (
              <ClickableText
                onPress={this.registerFreeAddress}
                icon={<Fontello name="register-new-fio-icon" color={theme.iconTappable} size={theme.rem(1)} />}
                label={lstrings.fio_address_reg_free}
              />
            ) : null}
            {this.renderErrorMessage()}
            <View style={styles.bottomSpace} />
          </View>
        </ScrollView>
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    paddingTop: theme.rem(0.5)
  },
  view: {
    position: 'relative'
  },
  createWalletPromptArea: {
    paddingHorizontal: theme.rem(1),
    paddingBottom: theme.rem(1)
  },
  instructionalText: {
    fontSize: theme.rem(1),
    color: theme.primaryText
  },
  handleRequirementsText: {
    fontSize: theme.rem(1),
    textAlign: 'left',
    color: theme.secondaryText
  },
  buttons: {
    marginTop: theme.rem(1.5),
    paddingHorizontal: theme.rem(1)
  },
  error: {
    flex: 1,
    margin: theme.rem(1)
  },
  addressTileBody: {
    flexDirection: 'row'
  },
  fioAddressName: {
    marginRight: theme.rem(1)
  },
  header: {
    flexDirection: 'row-reverse',
    justifyContent: 'flex-end',
    marginRight: theme.rem(1),
    marginTop: theme.rem(0.5)
  },
  image: {
    height: theme.rem(1.5),
    width: theme.rem(1.65),
    marginRight: theme.rem(0.5)
  },
  title: {
    paddingTop: theme.rem(1.5)
  },
  bottomSpace: {
    paddingBottom: theme.rem(30)
  },
  selectWalletBlock: {
    marginTop: theme.rem(3),
    paddingHorizontal: theme.rem(1.25),
    paddingBottom: theme.rem(0.75),
    backgroundColor: theme.tileBackground
  },
  selectWalletBtn: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: theme.rem(1),
    paddingVertical: theme.rem(0.5),
    paddingHorizontal: theme.rem(0.25),
    backgroundColor: theme.primaryButton[0]
  },
  domain: {
    marginTop: theme.rem(1.5),
    marginLeft: theme.rem(0.25),
    paddingHorizontal: theme.rem(0.75),
    paddingVertical: theme.rem(0.25),
    borderRadius: theme.rem(0.25),
    borderColor: theme.primaryButton[0],
    borderWidth: theme.rem(0.125)
  },
  domainText: {
    color: theme.primaryText,
    fontSize: theme.rem(1)
  },
  domainListRowName: {
    flex: 1,
    fontSize: theme.rem(1),
    color: theme.secondaryText
  },
  domainListRowContainerTop: {
    height: 'auto',
    paddingLeft: theme.rem(0.75),
    paddingRight: theme.rem(0.75),
    paddingVertical: theme.rem(0.75)
  },
  errorMessage: {
    color: theme.dangerText
  },
  muted: {
    color: theme.deactivatedText
  }
}))

const typeHack: any = {}

export const FioAddressRegisterScene = connect<StateProps, DispatchProps, OwnProps>(
  state => ({
    fioWallets: state.ui.wallets.fioWallets,
    fioPlugin: state.core.account.currencyConfig.fio ?? typeHack,
    isConnected: state.network.isConnected
  }),
  dispatch => ({
    async createFioWallet() {
      return await dispatch(createFioWallet())
    }
  })
)(withTheme(FioAddressRegister))
