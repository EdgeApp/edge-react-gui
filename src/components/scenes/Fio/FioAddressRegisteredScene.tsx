import * as React from 'react'
import { Image, View } from 'react-native'

import { formatDate } from '../../../locales/intl'
import { lstrings } from '../../../locales/strings'
import { EdgeSceneProps } from '../../../types/routerTypes'
import { SceneWrapper } from '../../common/SceneWrapper'
import { FormattedText } from '../../legacy/FormattedText/FormattedText.ui'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../../services/ThemeContext'
import { MainButton } from '../../themed/MainButton'

interface OwnProps extends EdgeSceneProps<'fioAddressRegisterSuccess'> {}

type Props = OwnProps & ThemeProps

export class FioAddressRegistered extends React.Component<Props> {
  renderExpDate() {
    const { theme, route } = this.props
    const { expiration } = route.params
    if (expiration != null) {
      const styles = getStyles(theme)

      return (
        <FormattedText style={styles.text}>
          {`${lstrings.fio_address_details_screen_expires} `}
          {formatDate(new Date(expiration))}
        </FormattedText>
      )
    }

    return null
  }

  render() {
    const { theme, navigation, route } = this.props
    const { fioName } = route.params
    const styles = getStyles(theme)

    return (
      <SceneWrapper background="theme">
        <View style={styles.view}>
          <View style={styles.texts}>
            <View style={styles.image}>
              <Image source={theme.fioAddressLogo} />
            </View>
            <FormattedText style={styles.text}>{lstrings.fio_address_details_screen_registered}</FormattedText>
            <FormattedText style={styles.title}>{fioName}</FormattedText>
            {this.renderExpDate()}
          </View>
          <MainButton marginRem={2} onPress={() => navigation.navigate('fioAddressList', {})} label={lstrings.title_fio_names} />
        </View>
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  view: {
    flex: 2,
    flexDirection: 'column',
    justifyContent: 'space-between',
    paddingHorizontal: theme.rem(1)
  },
  texts: {
    flex: 1,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: theme.rem(2.5)
  },
  text: {
    color: theme.primaryText,
    fontSize: theme.rem(1)
  },
  image: {
    marginLeft: theme.rem(0.25),
    marginBottom: theme.rem(3)
  },
  title: {
    fontSize: theme.rem(1.75),
    color: theme.primaryText,
    marginTop: theme.rem(1.25),
    marginBottom: theme.rem(0.75)
  },
  expiration: {
    fontSize: theme.rem(0.75),
    color: theme.primaryText,
    textAlign: 'center',
    marginTop: theme.rem(-0.5),
    paddingBottom: theme.rem(0.75)
  }
}))

export const FioAddressRegisteredScene = withTheme(FioAddressRegistered)
