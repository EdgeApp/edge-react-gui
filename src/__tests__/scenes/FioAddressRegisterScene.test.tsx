import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { FioAddressRegister } from '../../components/scenes/Fio/FioAddressRegisterScene'
import { getTheme } from '../../components/services/ThemeContext'
import { fakeSceneProps } from '../../util/fake/fakeSceneProps'

describe('FioAddressRegister', () => {
  it('should render with loading props', () => {
    const renderer = createRenderer()

    const fakeWallet: any = {
      currencyCode: 'FIO',
      nativeAmount: '100',
      networkFee: '1',
      blockHeight: 34,
      date: 220322,
      txid: '0x34346463',
      signedTx: '0xdgs3442',
      ourReceiveAddresses: ['FioAddress']
    }

    const actual = renderer.render(
      <FioAddressRegister
        {...fakeSceneProps('fioAddressRegister', {})}
        fioWallets={[fakeWallet]}
        fioPlugin={
          {
            currencyInfo: 'FIO plugin'
          } as any
        }
        isConnected
        createFioWallet={async () => 'myFio@fio' as any}
        theme={getTheme()}
      />
    )

    expect(actual).toMatchSnapshot()
  })
})
