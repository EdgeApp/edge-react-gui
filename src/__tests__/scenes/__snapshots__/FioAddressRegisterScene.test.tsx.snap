// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FioAddressRegister should render with loading props 1`] = `
<SceneWrapper
  background="theme"
>
  <Memo(SceneHeaderComponent)
    style={
      Object {
        "flexDirection": "row-reverse",
        "justifyContent": "flex-end",
        "marginRight": 22,
        "marginTop": 11,
      }
    }
    title="Register FIO Crypto Handle"
  >
    <Image
      resizeMode="cover"
      source={
        Object {
          "testUri": "../../../src/assets/images/details_fioAddress.png",
        }
      }
      style={
        Object {
          "height": 34,
          "marginRight": 11,
          "width": 37,
        }
      }
    />
  </Memo(SceneHeaderComponent)>
  <ScrollView>
    <View
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "paddingBottom": 22,
              "paddingHorizontal": 22,
            },
            Object {
              "paddingTop": 34,
            },
          ]
        }
      >
        <WithTheme(EdgeTextComponent)
          numberOfLines={2}
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 22,
            }
          }
        >
          Send and Receive with an easy to remember FIO Crypto Handle!
        </WithTheme(EdgeTextComponent)>
      </View>
      <View
        style={
          Object {
            "paddingBottom": 22,
            "paddingHorizontal": 22,
          }
        }
      >
        <WithTheme(EdgeTextComponent)
          numberOfLines={3}
          style={
            Object {
              "color": "#A4C7DF",
              "fontSize": 22,
              "textAlign": "left",
            }
          }
        >
          • Works across all tokens and coins
• Secured from your wallet
• Completely decentralized
        </WithTheme(EdgeTextComponent)>
      </View>
      <View
        style={
          Object {
            "paddingBottom": 22,
            "paddingHorizontal": 22,
          }
        }
      >
        <WithTheme(EdgeTextComponent)
          numberOfLines={5}
          style={
            Object {
              "color": "#A4C7DF",
              "fontSize": 22,
              "textAlign": "left",
            }
          }
        >
          Similar to website names, your FIO Crypto Handle can be used to send you tokens/coins without ever having to see or send your public key.
        </WithTheme(EdgeTextComponent)>
      </View>
      <View
        onLayout={[Function]}
      >
        <WithTheme(TileComponent)
          onPress={[Function]}
          title="Choose a FIO Crypto Handle"
          type="editable"
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <WithTheme(EdgeTextComponent)
              style={
                Object {
                  "color": "#87939E",
                }
              }
            >
              Your Crypto Handle name here
            </WithTheme(EdgeTextComponent)>
          </View>
        </WithTheme(TileComponent)>
        <WithTheme(TileComponent)
          body="Loading…"
          onPress={[Function]}
          title="Choose FIO Domain for Your Crypto Handle"
          type="touchable"
        />
      </View>
      <WithTheme(FormErrorComponent)
        isVisible={false}
        style={
          Object {
            "flex": 1,
            "margin": 22,
          }
        }
      >
        
      </WithTheme(FormErrorComponent)>
      <View
        style={
          Object {
            "paddingBottom": 674,
          }
        }
      />
    </View>
  </ScrollView>
</SceneWrapper>
`;
